tools/generate.py

```diff
@@ -1,276 +1,404 @@
+                    for i in range(len(text) - n + 1):
+                        ngram = text[i : i + n]
+                        lang_ngram_cnt[lang][ngram] += 1
+
+                    # when buffer is full, save it into table and empty it
+                    tot_ngrams = sum(len(v) for v in lang_ngram_cnt.values())
+                    if tot_ngrams >= buffer_length:
+                        for lang, ngram_hits in lang_ngram_cnt.items():
+                            self._upsert_ngram_hits(
+                                ngram_hits, lang, table_name
                             )
-            conn.commit()
+                        lang_ngram_cnt = defaultdict(lambda: defaultdict(int))
+
+            self._print_status(line_id, ngram_size=n, force=True)
+
+            # move remaining ngram hits from memory to database tables
+            for lang, ngram_hits in lang_ngram_cnt.items():
+                self._upsert_ngram_hits(ngram_hits, lang, table_name)
+
+            print(" done")
+
+        # save users contribution scores
+        self._insert_user_scores(user_lang_score)
+
+    def _init_db(self) -> None:
+        """Open the database connection and initialize the tables"""
+
+        with sqlite3.connect(self._fp) as conn:
+            # create a table for each n-gram type
+            for n in range(MIN_NGRAM_SIZE, MAX_NGRAM_SIZE + 1):
+                conn.execute(
+                    f"""
+                    CREATE TABLE IF NOT EXISTS grams{n} (
+                      'gram' text not null,
+                      'lang'text not null,
+                      'hit'  int not null,
+                    PRIMARY KEY("gram","lang")
+                    );
+                    """
+                )
+            # create a table dedicated to users contribution scores
+            conn.execute(
+                """
+                CREATE TABLE IF NOT EXISTS users_langs (
+                  'user' text not null,
+                  'lang' text not null,
+                  'total' int not null default 0
+                );
+                """
+            )
+
+    def _upsert_ngram_hits(
+        self, ngram_hits: dict, lang: str, table_name: str
+    ) -> None:
+        """Update n-gram hit count values in this table"""
+
+        with sqlite3.connect(self._fp) as conn:
+            # some optimization to make connection faster
+            conn.execute("PRAGMA journal_mode=MEMORY;")
+            conn.execute("PRAGMA temp_store=MEMORY;")
+
+            for ngram, hits in ngram_hits.items():
+                ngram = ngram.replace("'", "''")  # escape single quotes
+                conn.execute(
+                    f"""
+                    INSERT INTO {table_name} 
+                      VALUES ('{ngram}', '{lang}', {hits}) 
+                    ON CONFLICT (gram, lang) 
+                      DO UPDATE SET hit = hit + {hits};
+                    """
+                )
```

Comment:
No amount of replace() or other escaping tricks will truly protect you from SQL injections. Why not doing it the same way as the previous implementation, using `?` and passing the variables ngram and lang as second parameter of `conn.execute()`?